x-pack/plugins/infra/public/containers/logs/log_entries/index.ts

```diff
@@ -191,46 +262,72 @@ const useFetchEntriesEffect = (
 
   const fetchNewerEntries = useCallback(
     throttle(() => runFetchMoreEntriesRequest(ShouldFetchMoreEntries.After), 500),
-    [props, state.entriesEnd]
+    [props, state.bottomCursor]
```

Comment:
No. This callback only fetches entries at the bottom (`runFetchMoreEntriesRequest(ShouldFetchMoreEntries.After)`). It should not be affected by the top edge.

Reply:
But the function `runFetchMoreEntriesRequest` closes over `state.topCursor`, so it might use an outdated value. Not including a complete dependency list is considered dangerous because it is not robust to modification over time, even if it is "not buggy" at the time of writing (see https://overreacted.io/a-complete-guide-to-useeffect/#dont-lie-to-react-about-dependencies). I found quite a few related bugs while authoring #50979. I'll share my notes from that effort on Slack, but tl;dr is that there are ways to express the desired semantics in a safer and more explicit way.

One example would be

https://github.com/elastic/kibana/blob/29eda202376d32d3315d040491dbc006ce6c82a7/x-pack/plugins/infra/public/utils/use_url_state.ts#L74-L83

`usePrevious` is also often useful for conditionally triggering effects on changes (https://streamich.github.io/react-use/?path=/story/state-useprevious--docs):

```typescript
// assuming `value` is read from the props
const previousValue = usePrevious(value);

const runSideEffect = useCallback(() => { /* ... */}, [/* ... */])

useEffect(() => {
  if (value !== previousValue) {
    // This only runs when `value` changes no matter what `runSideEffect` closes over.
    // But it always uses the up-to-date reference to the function.
    runSideEffect();
  }
}, [previousValue, value, runSideEffect])
```

Reply:
> But the function runFetchMoreEntriesRequest closes over state.topCursor, so it might use an outdated value.

I don't understand why. That function is not a hook in itself (it's not wrapped under `useMemo`, `useCallback`...). It gets recreated on every call to `useLogEntriesState` so it will always have the latest value of `state` no matter what.

If we add `state.topCursor` to the dependencies the effect will add entries to the bottom when `state.topCursor` changes.

Am I missing something obvious?

Reply:
`useCallback` only uses the new function when the values in the dependency list change. So `runFetchMoreEntriesRequest` can be recreated all it wants (which is not ideal in itself, thus the proposal to `useCallback` that function in the first place), `useCallback` will return the old function as `fetchNewerEntries` unless the dependencies change.

Reply:
> If we add `state.topCursor` to the dependencies the effect will add entries to the bottom when `state.topCursor` changes.

Indeed, but the dependency list is not the way to prevent that, because it is supposed to model the lexical data flow and not the business logic. The `usePrevious`/`if` formulation I proposed in another comment would be a clean solution.

Reply:
> The usePrevious/if formulation I proposed in another comment would be a clean solution.

👍 Let's do that when we refactor the hooks in https://github.com/elastic/kibana/issues/59875